mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/Insert.java

```diff
@@ -125,4 +126,11 @@
    * the rules of {@link Long#parseLong(String)}), the mapper will issue a compile-time warning.
    */
   String timestamp() default "";
+
+  /**
+   * The nullSavingStrategy to use in the generated INSERT query.
+   *
+   * <p>The default value for @Insert is {@link NullSavingStrategy#SET_TO_NULL}
+   */
+  NullSavingStrategy nullSavingStrategy() default NullSavingStrategy.SET_TO_NULL;
```

Comment:
I understand that this is effectively needed because protocol V3 does not support leaving values unset, but this being the default was a truly sharp edge for the mapper in 3.x.

I know that C* 2.1 is still very popular, but I wonder if there's anything we should do to help mitigate our users dumping a ton of tombstones into their tables.

One possibility is that we could make `NullSavingStrategy.DO_NOT_SET` the default, and if is protocol V3 is selected, throw a preemptive runtime error earlier with a descriptive error.  I think this is better than adding tombstones unknowingly, and as more users embrace C 3.0+ this becomes a non-issue.   This wasn't done in the 3.x mapper because we added this capability after the mapper was originally released and we didn't want to change default behavior, but now that we're doing a major revision, changing the default is worth considering.

Thoughts?

Reply:
I agree with having 3.0-friendly defaults, and throwing early for lower versions (probably from the DAO's `initAsync`).

I think the default should depend on the annotation: for `@Update`, `DO_NOT_SET` feels intuitive for the partial update behavior, but with `@Insert`/`@SetEntity` I think I would expect everything to be saved/bound so `SET_TO_NULL`. `@Query` is less obvious as the query can be anything, I would go with `SET_TO_NULL` too.

Reply:
> for @Update, DO_NOT_SET feels intuitive for the partial update behavior

Ah, I did not realize `@Update` was implemented in this way.

> with @Insert/@SetEntity I  think I would expect everything to be saved/bound so SET_TO_NULL

I don't think we should treat the null setting behavior based on the kind of mutation operation differently as I think in either case we are being opinionated about whether the user intends to apply a tombstone or not when really we can't be sure.

One could even argue that setting to null on update makes more sense than insert as creating a tombstone is meaningful in that its a marker that can mean something is  deleted which is important when merging cells and determining if replicas are consistent.  Not creating a tombstone does not have that impact, which is why you could argue that inserts shouldn't create them.

Reply:
- Concerning catching exception add adding custom: 
 don't you think that this:
`Unset value at index 0. If you want this value to be null, please set it to null explicitly.`
is self-explanatory? (I've added test to capture that for 2.1:
https://github.com/datastax/java-driver/pull/1241/commits/d1042844209bbad29945e2dbcbd7aa7c0d135e11#diff-f21dbbd50b1f380dcadfd0e026c9e219R69)
Also as @tolbertam pointed out it is generated at the client side so there is no roundtrip involved.

- Concerning defaults - I agree with @tolbertam that it is better to have the same default for all cases.

>I don't think we should treat the null setting behavior based on the kind of mutation operation differently as I think in either case we are being opinionated about whether the user intends to apply a tombstone or not when really we can't be sure.

Setting it to `DO_NOT_SET` for `@Update` is our guess and may surprise users when for example they will choose to use firstly `@Query("UPDATE a ...)` and then they will decide to migrate `@Update`. In that case, the null saving strategy will change to `SET_TO_NULL` and they may be surprised by that fact. 
So maybe we should pick one default and use for all operations? (`@Query`, `@SetEntity`, `@Update` and `@Insert`)?
Another aspect is that I think we should instruct our users to set that nullSavingStrategy parameter explicitly - it will enforce them to make a conscious decision and read about those strategies.

Reply:
> catching exception

As discussed privately, it would be nice to generate a one-time check of the protocol version in `DAO.initAsync` if any method in the DAO interface uses `DO_NOT_SET`. You can do it with a callback `DaoImplementationSharedCode#addUnsetCapabilityCheck`.

> defaults

Yeah Andy convinced me. I was thinking of `Insert` in the case where the row does not exist yet, but then the strategy doesn't matter because the column will be null whether it was set or not.
Forcing the user to declare the strategy every time seems a bit harsh, so let's default to `DO_NOT_SET` everywhere.

Reply:
1. I've implemented custom one-time check where `V3` and `DO_NOT_SET`: https://github.com/datastax/java-driver/pull/1241/commits/ca4a9a3c9b1d08f1e642ca62f3da10f7dd731f50
I think that we don't need to add this method: `DaoImplementationSharedCode#addUnsetCapabilityCheck` to `DaoImplementationSharedCode` interface. It is enough to keep it private in `DaoImplementationGenerator`
2.1. I've changed defaults to DO_NOT_SET.

Reply:
> It is enough to keep it private

👍
You implemented differently than what I had in mind, indeed no need to make it an interface method.